Add the presto router

ChunxuTang opened the pull request:
This PR adds a presto router service in front of presto clusters, helping routing requests to presto clusters. The router makes it easier to scale out presto clusters with an aggregated Presto-styled UI and APIs.

Co-Authored-By: Hao Luo <hluo@twitter.com>
Co-Authored-By: Beinan Wang <beinanw@twitter.com>

Test plan - Added tests

```
== RELEASE NOTES ==

General Changes
* Add a router in front of presto clusters. The router is deployed as a standalone service with a presto-styled UI. 

```

ChunxuTang commented:
An example of the aggregated UI is shown below. It is served by the presto router.

![router-ui](https://user-images.githubusercontent.com/8018679/178162903-3bdc3477-f152-4e80-95f5-70fef1d03702.jpg)

ChunxuTang commented:
Hi @zhenxiao and @beinan, would you mind reviewing this new module? Thanks!

rongrong commented on presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterManager.java:
```diff
@@ -0,0 +1,117 @@
+        return Optional.ofNullable(candidates.get(RANDOM.nextInt(candidates.size())));
+    }
+
+    public Optional<String> matchGroup(RequestInfo requestInfo)
```
Does this need to be public?

ChunxuTang replied:
Thanks for pointing that out! I updated it to `private`.

rongrong commented on presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterStatusTracker.java:
```diff
@@ -0,0 +1,146 @@
+    @PostConstruct
+    public void startPollingQueryInfo()
+    {
+        clusterManager.getAllClusters().stream()
```
nits: `stream()` is probably not needed

ChunxuTang replied:
Yep. I removed the unnecessary `stream()` from the code (there's another place I used it redundantly and I also removed it from there).

rongrong commented on presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterStatusTracker.java:
```diff
@@ -0,0 +1,146 @@
+
+    private void pollQueryInfos()
+    {
+        Set<URI> allClusters = newHashSet(clusterManager.getAllClusters());
```
nits: Prefer `ImmutableSet`

ChunxuTang replied:
Nice catch! I updated the usage of `Set` to `ImmutableSet`.

rongrong reviewed (approved):
Thanks a lot for contributing the project! Just some nits.

rongrong commented:
Please rebase and fix the merge conflict.

7c00 reviewed (approved):
LGTM

ChunxuTang commented on presto-router/src/main/java/com/facebook/presto/router/RouterModule.java:
```diff
@@ -0,0 +1,67 @@
+{
+    private static final int IDLE_TIMEOUT_SECOND = 30;
+    private static final int REQUEST_TIMEOUT_SECOND = 10;
+    private static final String QUERY_TRACKER = "query-tracker";
```
Thanks for the suggestion! I've made the style here consistent.

ChunxuTang commented:
> Please extend the README with proper documentation on how to use it. Also consider add to official documentation.

Thanks a lot for the review!

I've rebased and fixed some merge conflicts. I've also updated the README file with necessary tutorials about how to use the module and the UI. Plan to send another PR to update the overall Presto doc for the new module.

ChunxuTang commented on presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterManager.java:
```diff
@@ -0,0 +1,117 @@
+
+public class ClusterManager
+{
+    private static final Random RANDOM = new Random();
```
Generally, there's only one cluster management instance as we usually don't have many Presto clusters running simultaneously (e.g., only ~10 at most). So we don't have to introduce thread-local variables for the current use cases. But I do agree that in the future when we need to handle a much higher number of Presto clusters, we'll need to introduce some concurrency mechanisms.

ChunxuTang commented on presto-router/README.txt:
```diff
@@ -0,0 +1 @@
+Presto Router
```
Thanks for the suggestion! I've added the instructions to launch this service in IntelliJ.

ChunxuTang commented on presto-router/src/main/java/com/facebook/presto/router/cluster/RemoteInfoFactory.java:
```diff
@@ -0,0 +1,50 @@
+public class RemoteInfoFactory
+{
+    private static final String QUERY_INFO = "/v1/query";
+    private static final String CLUSTER_INFO = "/v1/cluster";
```
That'll make it clearer! Thanks for the suggestion. I've updated the code.

ChunxuTang commented on presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterStatusTracker.java:
```diff
@@ -0,0 +1,146 @@
+
+        queryInfoUpdateExecutor.scheduleWithFixedDelay(() -> {
+            try {
+                pollQueryInfos();
```
As the router usually doesn't need to deal with a large num of Presto clusters (e.g. only ~10 at most), scheduling with a fixed delay generally looks sufficient. Maybe we can add a config to let users to customize the delay/jitter here.

ChunxuTang commented on presto-router/src/main/java/com/facebook/presto/router/cluster/RequestInfo.java:
```diff
@@ -0,0 +1,75 @@
+        return ImmutableList.copyOf(SPLITTER.split(nullToEmpty(servletRequest.getHeader(PRESTO_CLIENT_TAGS))));
+    }
+
+    private String parseHeader(HttpServletRequest servletRequest, String header)
```
Nice catch! Updated the code.

ChunxuTang commented on presto-router/src/main/java/com/facebook/presto/router/cluster/RequestInfo.java:
```diff
@@ -0,0 +1,75 @@
+        return clientTags;
+    }
+
+    private List<String> parseClientTags(HttpServletRequest servletRequest)
```
Thanks! Updated the code.

ChunxuTang commented:
@7c00 Thanks a lot for your help!

arunthirupathi commented:
The tests are flaky

https://github.com/prestodb/presto/runs/7398294709?check_suite_focus=true#step:7:21636

Error:  com.facebook.presto.router.TestClusterManager.testQuery  Time elapsed: 20.133 s  <<< FAILURE!
java.lang.AssertionError: expected [true] but found [false]
	at org.testng.Assert.fail(Assert.java:110)

2022-07-18T16:51:51.463-0500	INFO	TestNG-test=Surefire test-1	com.facebook.airlift.bootstrap.LifeCycleManager	Life cycle stopped.
	at com.facebook.presto.router.TestClusterManager.assertQueryState(TestClusterManager.java:149)
	at com.facebook.presto.router.TestClusterManager.testQuery(TestClusterManager.java:129)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)

ChunxuTang commented:
@arunthirupathi Thanks for figuring that out!
I've created https://github.com/prestodb/presto/pull/18063 to fix that. Would you mind having a quick review of it?

ChunxuTang commented:
FYI: The usage of the presto router is documented in https://github.com/prestodb/presto/pull/18118

highker commented:
Another flaky test fix: #18133
